src/os_cloud_nodelet.cpp

```diff
@@ -152,31 +188,112 @@ class OusterCloud : public nodelet::Nodelet {
             info.lidar_to_sensor_transform, sensor_frame, lidar_frame, msg_ts));
     }
 
-    void lidar_handler_sensor_time(const PacketMsg::ConstPtr& packet) {
-        if (!(*scan_batcher)(packet->buf.data(), ls)) return;
-        auto ts_v = ls.timestamp();
+    uint64_t impute_value(int last_scan_last_nonzero_idx,
+                          uint64_t last_scan_last_nonzero_value,
+                          int curr_scan_first_nonzero_idx,
+                          uint64_t curr_scan_first_nonzero_value,
+                          int scan_width) {
+        assert(scan_width + curr_scan_first_nonzero_idx >
+               last_scan_last_nonzero_idx);
+        double interpolated_value = linear_interpolate(
+            last_scan_last_nonzero_idx,
+            static_cast<double>(last_scan_last_nonzero_value),
```

Comment:
every `uin64_t` to `double` conversion seems loosing the precision here, because current timestamp values in nanoseconds take about 61 bits (`log2(1676586538159504896) = 60.54` and double mantissa is just 53 bits ...

the effect maybe not that big, but it can bite us very hard in some cases, I would recommend adding some unit tests with timestamp values in uint64_t nanosecs from the current UTC time and see what we really loosing and where it start breaking. Another factor is that measurement_ids are in 1000s ranges (from 0 to 4095 in some cases) and we multiply it to doubles that we receive from 61 bit unsigned int conversion ... ugh ...